### PR TITLE
ddl:  manage job lifecycle in worker

### PR DESCRIPTION
**Pull request by D3Hunter**
### What problem does this PR solve?

Issue Number: ref #53246

Problem Summary:

### What changed and how does it work?
- manage job life cycle inside worker
- move some methods for querying DDL related system tables to a separate pkg
- replace some `background` contexts to DDL worker ctx
- add context param to `begin/commit`, `rollback` is not changed as it doesn't return err and the main usages of ctx is to extract value

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by lance6716 — COMMENTED** [review]
14 / 24 files viewed

File: pkg/ddl/ddl_worker.go [hunk 1]
```diff
@@ -637,41 +637,25 @@ func cleanMDLInfo(pool *sess.Pool, job *model.Job, ec *clientv3.Client, ownerID
...
 	se := sess.NewSession(sctx)
 	se.GetSessionVars().SetDiskFullOpt(kvrpcpb.DiskFullOpt_AllowedOnAlmostFull)
-	_, err := se.Execute(context.Background(), sql, "delete-mdl-info")
+	_, err := se.Execute(s.schCtx, sql, "delete-mdl-info")
 	if err != nil {
 		logutil.DDLLogger().Warn("unexpected error when clean mdl info", zap.Int64("job ID", job.ID), zap.Error(err))
 		return
 	}
-	if cleanETCD && ec != nil {
+	// TODO do we need clean it for JobStateRollbackDone?
```
Comment (tangenta) [inline pkg/ddl/ddl_worker.go:649]: `JobStateRollbackDone` does trigger schema version changes for some DDLs, if we don't clean keys here, there may be a leak?

Comment (D3Hunter) [reply pkg/ddl/ddl_worker.go:649]: yes, that's why i want to add a JobStateRollbackSynced, so we can check it easily

File: pkg/ddl/ddl_running_jobs.go [hunk 2]
```diff
@@ -53,40 +53,38 @@ func (j *runningJobs) clear() {
...
 
-func (j *runningJobs) remove(job *model.Job) {
+func (j *runningJobs) remove(jobID int64, involvedSchemaInfos []model.InvolvingSchemaInfo) {
 	j.Lock()
 	defer j.Unlock()
-	delete(j.processingIDs, job.ID)
+	delete(j.processingIDs, jobID)
 	j.updateInternalRunningJobIDs()
 
-	if job.IsFinished() || job.IsSynced() {
```
Comment (D3Hunter) [inline pkg/ddl/ddl_running_jobs.go:81]: reorg jobs will check whether there are jobs of same table-id are `in-processing` in `getReorgJob`, so for a table-id there shouldn't be 2 jobs running concurrently, so it's ok to delete those unfinishedXXX all the time

File: pkg/infoschema/test/clustertablestest/cluster_tables_test.go [hunk 3]
```diff
@@ -1776,14 +1777,14 @@ func TestMDLViewIDConflict(t *testing.T) {
 	wg := &sync.WaitGroup{}
 	wg.Add(1)
 	go func() {
-		ddlTK1.MustExec("ALTER TABLE t ADD COLUMN b INT;")
+		ddlTK1.MustExec("ALTER TABLE t ADD index(a);")
```
Comment (D3Hunter) [inline pkg/infoschema/test/clustertablestest/cluster_tables_test.go:1780]: now the routine will hold worker until it exist, so we change to reorg job to make this test work

**Comment by D3Hunter** [comment]
@lance6716 @tangenta test pass, ptal

File: pkg/ddl/reorg.go [hunk 4]
```diff
@@ -892,13 +893,13 @@ func (r *reorgInfo) UpdateReorgMeta(startKey kv.Key, pool *sess.Pool) (err error
 	defer pool.Put(sctx)
 
 	se := sess.NewSession(sctx)
-	err = se.Begin()
+	err = se.Begin(context.Background())
```
Comment (lance6716) [inline pkg/ddl/reorg.go:896]: maybe use r.d.ctx

**Review by lance6716 — APPROVED** [review]

File: pkg/ddl/reorg.go (hunk 4 above)
Comment (D3Hunter) [reply pkg/ddl/reorg.go:896]: that's no much difference with background, it's only cancelled at shutdown. we should refactor to use job worker  ctx or disttask ctx later

File: pkg/ddl/ddl_running_jobs.go (hunk 2 above)
Comment (tangenta) [reply pkg/ddl/ddl_running_jobs.go:81]: @D3Hunter Is it not necessary to distinguish `processingIDs` from `unfinishedIDs`?

File: pkg/ddl/job_table.go [hunk 5]
```diff
@@ -507,93 +510,123 @@ func (d *ddl) delivery2LocalWorker(pool *workerPool, task *limitJobTask) {
...
+	metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Inc()
+	s.wg.RunWithLog(func() {
 		defer func() {
+			pool.put(wk)
 			failpoint.InjectCall("afterDelivery2Worker", job)
-			s.runningJobs.remove(job)
+			s.runningJobs.remove(jobID, involvedSchemaInfos)
 			asyncNotify(s.ddlJobNotifyCh)
 			metrics.DDLRunningJobCount.WithLabelValues(pool.tp().String()).Dec()
 			if wk.ctx.Err() != nil && ingest.LitBackCtxMgr != nil {
```
Comment (tangenta) [inline pkg/ddl/job_table.go:523]: At this point of time, `wk` is put back into the pool and may be get by someone else. But it is still used by here, will there be any problems?

Comment (D3Hunter) [reply pkg/ddl/job_table.go:523]: wk.ctx is same as scheduler.schCtx now, it should be ok，but let's move `push back` after last usage

Comment (D3Hunter) [reply pkg/ddl/job_table.go]: https://github.com/pingcap/tidb/pull/53747/commits/1e89963a3403428fb7df4ac91fa6eb889e91ac77

File: pkg/ddl/ddl_running_jobs.go (hunk 2 above)
Comment (D3Hunter) [reply pkg/ddl/ddl_running_jobs.go:81]: like what we have discussed, `runningJobs.remove` will only called when the job is in final state like synced/paused/cancelled or it's done by another owner, before that another job operating the same table-id or name should be blocked

**Pull request merged**